src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioSuppressionFixService.cs

```diff
@@ -117,50 +122,109 @@ public bool RemoveSuppressions(bool selectedErrorListEntriesOnly, IVsHierarchy p
             }
         }
 
+        private async Task<ImmutableArray<DiagnosticData>> GetAllDiagnosticsAsync(Func<Project, bool> shouldFixInProject, CancellationToken cancellationToken)
+        {
+            var builder = ImmutableArray.CreateBuilder<DiagnosticData>();
+            var solution = _workspace.CurrentSolution;
+            foreach (var project in solution.Projects)
+            {
+                if (shouldFixInProject(project))
+                {
+                    var diagnostics = await _diagnosticService.GetDiagnosticsAsync(solution, project.Id, includeSuppressedDiagnostics: true, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    builder.AddRange(diagnostics.Where(d => d.Severity != DiagnosticSeverity.Hidden));
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static string GetFixTitle(bool isAddSuppression)
+        {
+            return isAddSuppression ? ServicesVSResources.SuppressMultipleOccurrences : ServicesVSResources.RemoveSuppressMultipleOccurrences;
+        }
+
+        private static string GetWaitDialogMessage(bool isAddSuppression)
+        {
+            return isAddSuppression ? ServicesVSResources.ComputingSuppressionFix : ServicesVSResources.ComputingRemoveSuppressionFix;
+        }
+
+        private IEnumerable<DiagnosticData> GetDiagnosticsToFix(Func<Project, bool> shouldFixInProject, bool selectedEntriesOnly, bool isAddSuppression)
+        {
+            var diagnosticsToFix = SpecializedCollections.EmptyEnumerable<DiagnosticData>();
+            Action<CancellationToken> computeDiagnosticsToFix = cancellationToken =>
+            {
+                // If we are fixing selected diagnostics in error list, then get the diagnostics from error list entry snapshots.
+                // Otherwise, get all diagnostics from the diagnostic service.
+                var diagnosticsToFixTask = selectedEntriesOnly ?
+                    _suppressionStateService.GetSelectedItemsAsync(isAddSuppression, cancellationToken) :
+                    GetAllDiagnosticsAsync(shouldFixInProject, cancellationToken);
+
+                diagnosticsToFix = diagnosticsToFixTask.WaitAndGetResult(cancellationToken);
```

Comment:
Making this async causes the code to become quite complicated with lot of duplication, reason being we need to realize the diagnostics at multiple points and we want to do that within a cancellable dialog... Let me come talk to you in person about this later today - I will cleanup the PR subsequently